python/ray/data/preprocessors/encoder.py

```diff
@@ -324,7 +325,9 @@ def _transform_pandas(self, df: pd.DataFrame):
         _validate_df(df, *self.columns)
```

Comment:
if the problem is existing encoder cannot handle arrow data format that leads to numpy input format, shouldn't we implement `_transform_numpy` for this preprocessor instead ? Then in our existing preprocessor / batch mapper / predictor code, we will be able to locally pick the numpy path based on input format and preprocessor implementations.

Enhancing `_transform_pandas` here still introduces copies of input data between pandas <> arrow

Reply:
> if the problem is existing encoder cannot handle arrow data format that leads to numpy input format, shouldn't we implement _transform_numpy for this preprocessor instead ?

I'm not sure.

I wouldn't say that the Arrow data format leads to a NumPy input format. It's still tabular data; it's just that lists are encoded as NumPy arrays. Since the data represents tabular data, I think `_transform_pandas` might be easier to implement.

```
>>> df = pd.DataFrame({
...        "movie": ["spam", "ham"],
...        "genre": [["action"], ["comedy", "romance"]]
... })
>>> type(df["genre"][0])
<class 'list'>
>>> df
  movie              genre
0  spam           [action]
1   ham  [comedy, romance]
```

```
>>> df = pa.table(df).to_pandas()
>>> type(df["genre"][0])
<class 'numpy.ndarray'>
>>> df
  movie              genre
0  spam           [action]
1   ham  [comedy, romance]
```

> Enhancing _transform_pandas here still introduces copies of input data between pandas <> arrow

Which copy are you referring to? Converting an Arrow Table to a pandas DataFrame?

Reply:
Agree with @bveeramani here- the problem is that the actual item of a particular row is a numpy array and not a list. Even with numpy path, this would still be a problem (the input would look like this: `{"column": np.array([np.array(["A", "B"])])}`). It's the inner numpy array that's troublesome.

We can still implement a `_transform_numpy` path, but as @bveeramani mentioned, given that this preprocessor is meant for tabular data and not tensors, not sure how much perf value we will get from this.

Regardless, this change is needed to make the pandas path work well.

Reply:
i see, from description of the PR, where does `it converts Arrow blocks to pandas DataFrames` happen ? I was under the impression that you're suggesting original data format is arrow but we only have `_transform_pandas` in this component, in this case adding `_transform_numpy` will be preferred.

But if original data format is tabular, enhancing `_transform_pandas` is the right way to go.  

As a follow up I would like to understand the workload you're optimizing for e2e, since the snippet you had above
```python
df = pa.table(df).to_pandas()
```
does't have to call `to_pandas` if we don't need to, right ? What use cases we will start with an arrow dataset ?

Reply:
>where does it converts Arrow blocks to pandas DataFrames happen ?

I think `map_batches`

https://github.com/ray-project/ray/blob/892b4f0de2c2311fbb6f2a3989fc4e22cb8994b2/python/ray/data/dataset.py#L596-L598

eventually calls `pa.Table.to_pandas`: 

https://github.com/ray-project/ray/blob/892b4f0de2c2311fbb6f2a3989fc4e22cb8994b2/python/ray/data/_internal/arrow_block.py#L222-L229

>  I was under the impression that you're suggesting original data format is arrow but we only have _transform_pandas in this component, in this case adding _transform_numpy will be preferred.

I think this is the case? The block type is Arrow, and `MultiHotEncoder` only implements `_transform_pandas`. That said, the dataset represents tabular data.

> As a follow up I would like to understand the workload you're optimizing for e2e, since the snippet you had above

Let's say you have a dataset of Tweets, and each Tweet has a "Tags" column that represents a list of tags. In this case,  `MultiHotEncoder` doesn't work:
```
dataset = ray.data.read_parquet("s3://tweets") # Arrow dataset that represents tabular data
preprocessor = MultiHotEncoder(columns=["Tags"])
preprocessor.fit_transform(dataset) # This errors
```